File: editor-server/src/resolvers/dancer.ts
import {
  Resolver,
  Query,
  Mutation,
  Ctx,
  PubSub,
  Publisher,
  Arg,
} from "type-graphql";

import { Dancer } from "../../prisma/generated/type-graphql";
import {
  AddDancerInput,
  editDancerInput,
  deleteDancerInput,
} from "./inputs/dancer";
import { Topic } from "./subscriptions/topic";
import { DancerPayload, dancerMutation } from "./subscriptions/dancer";
import { DancerResponse } from "./response/dancerResponse";
import { initRedisControl, initRedisPosition } from "../utility";
import { TContext } from "../types/global";

@Resolver((of) => Dancer)
export class DancerResolver {
  @Query((returns) => [Dancer])
  async dancers(@Ctx() ctx: TContext) {
    const dancers = await ctx.prisma.dancer.findMany({
      include: { parts: true, positionData: true }
    });
    return dancers;
  }

  @Query((returns) => Dancer)
  async dancer(
    @Arg("dancerName") dancerName: string,
    @Ctx() ctx: TContext
  ) {
    const dancer = await ctx.prisma.dancer.findFirst({
      where: { name: dancerName },
      include: { parts: true, positionData: true }
    });
    if(!dancer) throw new Error("dancer not found");
    return dancer;
  }

  @Mutation((returns) => DancerResponse)
  async addDancer(
    @PubSub(Topic.Dancer) publish: Publisher<DancerPayload>,
    @Arg("dancer") newDancerData: AddDancerInput,
    @Ctx() ctx: TContext
  ) {
    const existDancer = await ctx.prisma.dancer.findFirst({
      where: { name: newDancerData.name },
    });
    if (!existDancer) {
      const allPositionFramesIds = await ctx.prisma.positionFrame.findMany({
        select: { id: true }
      });
      const newDancer = await ctx.prisma.dancer.create({
        data: {
          name: newDancerData.name,
          positionData: {
            create: allPositionFramesIds.map(({ id: frameId }) => ({
              frameId,
              x: 0,
              y: 0,
              z: 0,
            })),
          }
        }
      });
      await initRedisControl();
      await initRedisPosition();
      const payload: DancerPayload = {
        mutation: dancerMutation.CREATED,
        editBy: Number(ctx.userID),
        dancerData: newDancer,
      };
      await publish(payload);

      // save dancer
      return {dancerData:newDancer,  ok: true, msg: "dancer created" };
    }
    return {dancerData: existDancer, ok: false, msg: "dancer exists" };
  }

  @Mutation((returns) => DancerResponse)
  async editDancer(
    @PubSub(Topic.Dancer) publish: Publisher<DancerPayload>,
    @Arg("dancer") newDancerData: editDancerInput,
    @Ctx() ctx: TContext
  ) {
    const { id, name } = newDancerData;
    const newDancer = await ctx.prisma.dancer.update({
      where: { id },
      data: { name },
    });
    if (newDancer) {
      const payload: DancerPayload = {
        mutation: dancerMutation.UPDATED,
        editBy: Number(ctx.userID),
        dancerData: newDancer,
      };
      await publish(payload);
      return {dancerData: newDancer, ok: true, msg: "dancer updated" };
    }
    return { ok: false, msg: "dancer not found" };
  }

  @Mutation((returns) => DancerResponse)
  async deleteDancer(
    @PubSub(Topic.Dancer) publish: Publisher<DancerPayload>,
    @Arg("dancer") delDancerData: deleteDancerInput,
    @Ctx() ctx: TContext
  ) {
    const { id } = delDancerData;
    const dancer = await ctx.prisma.dancer.findFirst({
      where: { id },
      include: { parts: true, positionData: true }
    });
    if(!dancer) return { ok: false, msg: "dancer not found" };
    await ctx.prisma.dancer.delete({ where: { id }});
    await initRedisControl();
    await initRedisPosition();
    const payload: DancerPayload = {
      mutation: dancerMutation.DELETED,
      editBy: Number(ctx.userID),
    };
    await publish(payload);
    return { ok: true, msg: "dancer deleted" };
  }
}

File: editor-server/src/graphql/mutations/dancer.rs
//! Dancer mutation methods.
use crate::db::types::dancer::DancerData;
use crate::graphql::{
    subscriptions::dancer::{DancerMutationMode, DancerPayload},
    subscriptor::Subscriptor,
    types::dancer::Dancer,
};
use crate::types::global::UserContext;
use crate::utils::data::{init_redis_control, init_redis_position};

use async_graphql::{Context, InputObject, Object, Result as GQLResult, SimpleObject};

#[derive(InputObject, Default, Debug)]
pub struct DancerUpdateInput {
    pub name: String,
    pub id: i32,
}

#[derive(InputObject, Default, Debug)]
pub struct DancerCreateInput {
    pub name: String,
    pub model: String,
}

#[derive(InputObject, Default, Debug)]
pub struct DancerDeleteInput {
    pub id: i32,
}

#[derive(SimpleObject, Default, Debug)]
pub struct DancerMutationResponse {
    ok: bool,
    msg: String,
}

#[derive(Default)]
pub struct DancerMutation;

#[Object]
impl DancerMutation {
    #[allow(unused)]
    async fn add_dancer(
        &self,
        ctx: &Context<'_>,
        input: DancerCreateInput,
    ) -> GQLResult<DancerMutationResponse> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();
        let redis = clients.redis_client();

        tracing::info!("Mutation: addDancer");

        let dancer_name = input.name.clone();

        let dancer_result = sqlx::query_as!(
            DancerData,
            r#"
                SELECT * FROM Dancer WHERE name = ?;
            "#,
            &dancer_name
        )
        .fetch_one(mysql)
        .await;

        if let Ok(dancer) = dancer_result {
            return Ok(DancerMutationResponse {
                ok: false,
                msg: "Dancer already exists.".to_string(),
            });
        }

        let raw_model = sqlx::query!(
            r#"
                SELECT id FROM Model WHERE name = ?;
            "#,
            &input.model
        )
        .fetch_one(mysql)
        .await?;

        let dancer_id = sqlx::query!(
            r#"
                INSERT INTO Dancer (name, model_id) VALUES (?, ?);
            "#,
            &dancer_name,
            raw_model.id
        )
        .execute(mysql)
        .await?
        .last_insert_id() as i32;

        init_redis_control(mysql, redis).await?;
        init_redis_position(mysql, redis).await?;

        let dancer_payload = DancerPayload {
            mutation: DancerMutationMode::Created,
            dancer_data: Some(Dancer {
                id: dancer_id,
                name: dancer_name,
                parts: None,
                position_datas: None,
            }),
            edit_by: context.user_id,
        };

        Subscriptor::publish(dancer_payload);

        Ok(DancerMutationResponse {
            ok: true,
            msg: "Dancer added".to_string(),
        })
    }

    #[allow(unused)]
    async fn edit_dancer(
        &self,
        ctx: &Context<'_>,
        input: DancerUpdateInput,
    ) -> GQLResult<DancerMutationResponse> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();

        tracing::info!("Mutation: editDancer");

        let dancer_id = input.id;
        let dancer_name = input.name.clone();

        let raw_dancer = sqlx::query_as!(
            DancerData,
            r#"
                SELECT * FROM Dancer WHERE id = ?;
            "#,
            &dancer_id,
        )
        .fetch_one(mysql)
        .await;

        if raw_dancer.is_err() {
            return Ok(DancerMutationResponse {
                ok: false,
                msg: "Dancer not found.".to_string(),
            });
        }

        let dancer_id = match raw_dancer {
            Ok(dancer) => dancer,
            Err(_) => {
                return Ok(DancerMutationResponse {
                    ok: false,
                    msg: "Dancer not found.".to_string(),
                })
            }
        };

        sqlx::query!(
            r#"
                UPDATE Dancer SET name = ? WHERE id = ?;
            "#,
            &dancer_name,
            &dancer_id.id
        )
        .execute(mysql)
        .await?;

        let dancer_payload = DancerPayload {
            mutation: DancerMutationMode::Updated,
            dancer_data: Some(Dancer {
                id: input.id,
                name: input.name.clone(),
                parts: None,
                position_datas: None,
            }),
            edit_by: context.user_id,
        };

        Subscriptor::publish(dancer_payload);

        Ok(DancerMutationResponse {
            ok: true,
            msg: "Dancer updated".to_string(),
        })
    }

    #[allow(unused)]
    async fn delete_dancer(
        &self,
        ctx: &Context<'_>,
        input: DancerDeleteInput,
    ) -> GQLResult<DancerMutationResponse> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();
        let redis = clients.redis_client();

        tracing::info!("Mutation: deleteDancer");

        let dancer_id = input.id;

        let raw_dancer = sqlx::query_as!(
            DancerData,
            r#"
                SELECT * FROM Dancer WHERE id = ?;
            "#,
            &dancer_id
        )
        .fetch_one(mysql)
        .await;

        let _dancer = match raw_dancer {
            Ok(dancer) => dancer,
            Err(_) => {
                return Ok(DancerMutationResponse {
                    ok: false,
                    msg: "Dancer not found.".to_string(),
                })
            }
        };

        let _ = sqlx::query!(
            r#"
                DELETE FROM Dancer WHERE id = ?;
            "#,
            &dancer_id
        )
        .execute(mysql)
        .await?;

        let _ = init_redis_control(mysql, redis).await?;
        let _ = init_redis_position(mysql, redis).await?;

        let dancer_payload = DancerPayload {
            mutation: DancerMutationMode::Deleted,
            dancer_data: None,
            edit_by: context.user_id,
        };

        Subscriptor::publish(dancer_payload);

        Ok(DancerMutationResponse {
            ok: true,
            msg: "Dancer updated".to_string(),
        })
    }
}
